src/io/flutter/devtools/DevToolsManager.java

```diff
@@ -259,26 +260,49 @@ private void openBrowserImpl(String uri, String screen) {
       devToolsInstance.openBrowserAndConnect(uri, screen);
     }
     else {
-      @Nullable final OSProcessHandler handler =
-        isBazel(project) ? getProcessHandlerForBazel() : getProcessHandlerForPub();
-
-      if (handler != null) {
-        // TODO(devoncarew) Add a Task.Backgroundable here; "Starting devtools..."
-
-        // start the server
-        DevToolsInstance.startServer(handler, instance -> {
-          devToolsInstance = instance;
+      // For internal users, we can connect to the DevTools server started by flutter daemon. For external users, the flutter daemon has an
+      // older version of DevTools, so we launch the server using `pub global run` instead.
+      if (isBazel(project)) {
+        final Optional<FlutterApp> first =
+          FlutterApp.allFromProjectProcess(project).stream().filter((FlutterApp app) -> app.getProject() == project).findFirst();
+
+        if (!first.isPresent()) {
+          return;
+        }
 
-          devToolsInstance.openBrowserAndConnect(uri, screen);
-        }, () -> {
-          // Listen for closing; null out the devToolsInstance.
-          devToolsInstance = null;
+        first.get().serveDevTools().thenAccept((Pair<String, Integer> hostAndPort) -> {
+          if (hostAndPort == null) {
```

Comment:
at each async boundary we probably need to make sure the project hasn't been closed in whcih case we should quietly cancel the request.

Reply:
This means a check of whether someone has closed down the IntelliJ project? And we probably should also check if the app has been disconnected?